### DISCUSSION TITLE
Why React Hydration Error considered error?

### DISCUSSION THREAD

**Discussion by skorphil** [vercel/next.js#60918] [Help]
### Summary

Hi, i'm playing with RSC and use `const isClient = typeof window !== "undefined"` to check if my components are client or server. 
However, it returns error when i use it in `client` components: https://nextjs.org/docs/messages/react-hydration-error

```
<ServerComponent>
    <ClientComponent />
</ ServerComponent>
```

I wonder, why does it considered an error? Isn't it how client components should work? I mean:

1. We pre-render component
2. We use client-specific logic to complete the component (change it after pre-rendering)
3. We return full component

This is feels natural, that a client component is changing after it's being pre-rendered. How do I suppose to deal with situations when i need  a client component to perform client-side functions like `const isClient = typeof window !== "undefined"`?

**Comment by amineshon** [comment]
Hydration errors occur when React tries to attach event handlers and manage the state of a component that doesn't match the initial rendering. This mismatch can lead to unexpected behavior, such as missing or doubled event listeners, unhandled state changes, or even crashing the application.

**Reply by amineshon** [reply]
[Nextjs React Hydration Error](https://nextjs.org/docs/messages/react-hydration-error)

[Fix Next.js: "Text content does not match server-rendered HTML.](https://www.netlify.com/blog/fix-next-js-react-hydration-error/)"

**Reply by little077** [reply]
Perhaps there should be a small case to tell developers why such errors occur, and the problems that may arise after errors occur, so that developers can know the trade-offs, so that they can control whether strong fixes are needed?

**Comment by icyJoseph** [comment]
The problem and burden of hydration, is that it has to replay the server rendered frame, in order to apply event listeners where they ought to be. 

In some cases, where this happens, React fallsback, though it errors anyway, to full client side render. However, the issue remains that, it was not possible to walk the HTML, using the VDOM map, to apply event listeners where needed.
